File: cliphook.hs
{-# LANGUAGE OverloadedStrings #-}

import System.Process (readProcess)
import Control.Concurrent (threadDelay)
import Crypto.Hash.SHA256 (hash)
import qualified Data.ByteString.Char8 as BS
import Data.ByteString.Base16 (encode)
import Data.Time.Clock
import Data.Time.Format
import System.Locale (defaultTimeLocale)
import System.IO (appendFile)

logFile :: FilePath
logFile = "/home/dein_user/clipboard.log"

interval :: Int
interval = 2 * 1000000  -- Mikrosekunden (2 Sekunden)

-- Clipboard auslesen
getClipboard :: IO String
getClipboard = do
    result <- readProcess "wl-paste" ["-n"] ""
    return $ trim result
  where
    trim = reverse . dropWhile (`elem` "\n\r") . reverse

-- SHA256 Hash
sha256Hash :: String -> String
sha256Hash s = BS.unpack $ encode $ hash $ BS.pack s

-- Loggen
logClipboard :: String -> IO ()
logClipboard content = do
    now <- getCurrentTime
    let timestamp = formatTime defaultTimeLocale "%Y-%m-%d %H:%M:%S" now
    appendFile logFile $ "[" ++ timestamp ++ "]\n" ++ content ++ "\n\n"

-- Hauptloop
loop :: String -> IO ()
loop lastHash = do
    content <- getClipboard
    let currentHash = sha256Hash content
    if not (null content) && currentHash /= lastHash
        then do
            logClipboard content
            putStrLn $ "Neuer Clipboard-Inhalt geloggt: " ++ take 30 content ++ "..."
            loop currentHash
        else do
            threadDelay interval
            loop lastHash

main :: IO ()
main = loop ""

File: cliphook.py
#!/usr/bin/env python3
import subprocess
import time
import hashlib
from datetime import datetime

# Log-Datei
LOG_FILE = "/home/dein_user/clipboard.log"

# Intervall in Sekunden
INTERVAL = 2

def get_clipboard():
    """Liest das Wayland-Clipboard via wl-paste"""
    try:
        result = subprocess.run(["wl-paste", "-n"], capture_output=True, text=True)
        return result.stdout.strip()
    except FileNotFoundError:
        print("wl-paste nicht gefunden! Bitte installieren.")
        exit(1)

def hash_content(content):
    """Gibt einen Hash zurück, um Änderungen schnell zu erkennen"""
    return hashlib.sha256(content.encode()).hexdigest()

def log_clipboard(content):
    """Schreibt Inhalt mit Timestamp in die Log-Datei"""
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    with open(LOG_FILE, "a") as f:
        f.write(f"[{timestamp}]\n{content}\n\n")

def main():
    last_hash = None
    while True:
        content = get_clipboard()
        current_hash = hash_content(content)
        if current_hash != last_hash:
            if content:  # nur nicht-leere Inhalte loggen
                log_clipboard(content)
                print(f"Neuer Inhalt geloggt: {content[:30]}...")
            last_hash = current_hash
        time.sleep(INTERVAL)

if __name__ == "__main__":
    main()

